test/rulesets/SparseArrays/sparsematrix.jl

```diff
@@ -18,6 +18,51 @@ end
     test_rrule(SparseVector{Float32}, Float32.(v), rtol=1e-4)
 end
 
+# copied over from test/rulesets/LinearAlgebra/structured
+@testset "spdiagm" begin
+    @testset "without size" begin
+        M, N = 7, 9
+        s = (8, 8)
+        a, ā = randn(M), randn(M)
+        b, b̄ = randn(M), randn(M)
+        c, c̄ = randn(M - 1), randn(M - 1)
+        ȳ = randn(s)
+        ps = (0 => a, 1 => b, 0 => c)
+        y, back = rrule(spdiagm, ps...)
```

Comment:
does:
```
test_rrule(spdiagm, ps...)
```
not work for this?

I suspect it does now, and when the features were added to make it work the code in was not all updated

Reply:
No, it doesn't work...I'd tried that previously, but it fails with the stacktrace below. Something similar happens with `test_rrule(diagm, ps...)`, which I assume is why the more verbose/manual tests are still there in `test/rulesets/LinearAlgebra/structured.jl`. I wasn't able to quickly figure out where the error was coming from, so I just copied the same code over.

<details>
<summary>Stacktrace</summary>

```
test_rrule: spdiagm on Pair{Int64, Vector{Float64}},Pair{Int64, Vector{Float64}},Pair{Int64, Vector{Float64}}: Error During Test at C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:202
  Got exception outside of a @test
  TypeError: in new, expected Int64, got a value of type Float64
  Stacktrace:
    [1] macro expansion
      @ C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\to_vec.jl:0 [inlined]
    [2] _force_construct
      @ C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\to_vec.jl:27 [inlined]
    [3] (::FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}})(v::Vector{Float64})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\to_vec.jl:53
    [4] (::FiniteDifferences.var"#52#54"{Vector{Float64}})(x_back::FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, l::Int64, s::Int64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\to_vec.jl:245
    [5] map(f::Function, t1::Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}, t2::Tuple{Int64, Int64, 
Int64}, ts::Tuple{Int64, Int64, Int64})
      @ Base .\tuple.jl:318
    [6] Tuple_from_vec
      @ C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\to_vec.jl:244 [inlined]
    [7] (::FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}})(x::Vector{Float64})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:77
    [8] (::FiniteDifferences.var"#74#76"{Int64, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}})(ε::Float64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:20
    [9] newf
      @ C:\Users\sam.urmy\.julia\packages\StaticArrays\dTwvg\src\broadcast.jl:185 [inlined]
   [10] macro expansion
      @ C:\Users\sam.urmy\.julia\packages\StaticArrays\dTwvg\src\broadcast.jl:140 [inlined]
   [11] __broadcast
      @ C:\Users\sam.urmy\.julia\packages\StaticArrays\dTwvg\src\broadcast.jl:128 [inlined]
   [12] _broadcast
      @ C:\Users\sam.urmy\.julia\packages\StaticArrays\dTwvg\src\broadcast.jl:124 [inlined]
   [13] copy
      @ C:\Users\sam.urmy\.julia\packages\StaticArrays\dTwvg\src\broadcast.jl:62 [inlined]
   [14] materialize
      @ .\broadcast.jl:873 [inlined]
   [15] _eval_function(m::FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}, f::FiniteDifferences.var"#74#76"{Int64, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}}, x::Float64, step::Float64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\methods.jl:249
   [16] _estimate_magnitudes(m::FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}, f::FiniteDifferences.var"#74#76"{Int64, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}}, x::Float64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\methods.jl:378
   [17] estimate_step(m::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::FiniteDifferences.var"#74#76"{Int64, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}}, x::Float64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\methods.jl:365
   [18] (::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}})(f::FiniteDifferences.var"#74#76"{Int64, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}}, x::Float64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\methods.jl:193
   [19] #73
      @ C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:16 [inlined]
   [20] iterate
      @ .\generator.jl:47 [inlined]
   [21] _collect(c::Base.OneTo{Int64}, itr::Base.Generator{Base.OneTo{Int64}, FiniteDifferences.var"#73#75"{FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}}}, #unused#::Base.EltypeUnknown, isz::Base.HasShape{1})
      @ Base .\array.jl:802
   [22] collect_similar
      @ .\array.jl:711 [inlined]
   [23] map
      @ .\abstractarray.jl:3263 [inlined]
   [24] jacobian(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, x::Vector{Float64}; len::Nothing)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:15
   [25] jacobian(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::Function, x::Vector{Float64})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:9
   [26] _j′vp(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::Function, ȳ::Vector{Float64}, x::Vector{Float64})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:84
   [27] j′vp(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, ȳ::SparseMatrixCSC{Float64, Int64}, x::Tuple{Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:77
   [28] j′vp(::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, ::Function, ::SparseMatrixCSC{Float64, Int64}, ::Pair{Int64, Vector{Float64}}, ::Pair{Int64, Vector{Float64}}, ::Vararg{Pair{Int64, Vector{Float64}}})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:80
   [29] _make_j′vp_call(fdm::Any, f::Any, ȳ::Any, xs::Any, ignores::Any)
      @ ChainRulesTestUtils C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\finite_difference_calls.jl:51
   [30] macro expansion
      @ C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:233 [inlined]
   [31] macro expansion
      @ C:\Users\sam.urmy\.julia\juliaup\julia-1.9.3+0.x64.w64.mingw32\share\julia\stdlib\v1.9\Test\src\Test.jl:1498 [inlined]
   [32] test_rrule(::RuleConfig, ::Any, ::Any, ::Vararg{Any}; output_tangent::Any, check_thunked_output_tangent::Any, fdm::Any, rrule_f::Any, check_inferred::Bool, fkwargs::NamedTuple, rtol::Real, atol::Real, testset_name::Any, kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
      @ ChainRulesTestUtils C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:205
   [33] test_rrule(::RuleConfig, ::Any, ::Any, ::Vararg{Any})
      @ ChainRulesTestUtils C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:181
   [34] test_rrule(::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
      @ ChainRulesTestUtils C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:178
   [35] test_rrule(::Any, ::Any, ::Any, ::Vararg{Any})
      @ ChainRulesTestUtils C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:177
   [36] macro expansion
      @ C:\Users\sam.urmy\.julia\dev\ChainRules\test\rulesets\SparseArrays\sparsematrix.jl:31 [inlined]
   [37] macro expansion
      @ C:\Users\sam.urmy\.julia\juliaup\julia-1.9.3+0.x64.w64.mingw32\share\julia\stdlib\v1.9\Test\src\Test.jl:1498 [inlined]
   [38] macro expansion
      @ C:\Users\sam.urmy\.julia\dev\ChainRules\test\rulesets\SparseArrays\sparsematrix.jl:24 [inlined]
   [39] macro expansion
      @ C:\Users\sam.urmy\.julia\juliaup\julia-1.9.3+0.x64.w64.mingw32\share\julia\stdlib\v1.9\Test\src\Test.jl:1498 [inlined]
   [40] top-level scope
      @ C:\Users\sam.urmy\.julia\dev\ChainRules\test\rulesets\SparseArrays\sparsematrix.jl:23
   [41] include(fname::String)
      @ Base.MainInclude .\client.jl:478
   [42] macro expansion
      @ C:\Users\sam.urmy\.julia\dev\ChainRules\test\runtests.jl:86 [inlined]
   [43] macro expansion
      @ C:\Users\sam.urmy\.julia\juliaup\julia-1.9.3+0.x64.w64.mingw32\share\julia\stdlib\v1.9\Test\src\Test.jl:1498 [inlined]
   [44] top-level scope
      @ C:\Users\sam.urmy\.julia\dev\ChainRules\test\runtests.jl:50
   [45] include(fname::String)
      @ Base.MainInclude .\client.jl:478
   [46] top-level scope
      @ none:6
   [47] eval
      @ .\boot.jl:370 [inlined]
   [48] exec_options(opts::Base.JLOptions)
      @ Base .\client.jl:280
   [49] _start()
      @ Base .\client.jl:522

  caused by: InexactError: Int64(-0.01)
  Stacktrace:
    [1] Int64
      @ .\float.jl:900 [inlined]
    [2] convert
      @ .\number.jl:7 [inlined]
    [3] Pair
      @ .\Base.jl:140 [inlined]
    [4] (::FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}})(v::Vector{Float64})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\to_vec.jl:51
    [5] (::FiniteDifferences.var"#52#54"{Vector{Float64}})(x_back::FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, l::Int64, s::Int64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\to_vec.jl:245
    [6] map(f::Function, t1::Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}, t2::Tuple{Int64, Int64, 
Int64}, ts::Tuple{Int64, Int64, Int64})
      @ Base .\tuple.jl:318
    [7] Tuple_from_vec
      @ C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\to_vec.jl:244 [inlined]
    [8] (::FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}})(x::Vector{Float64})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:77
    [9] (::FiniteDifferences.var"#74#76"{Int64, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}})(ε::Float64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:20
   [10] newf
      @ C:\Users\sam.urmy\.julia\packages\StaticArrays\dTwvg\src\broadcast.jl:185 [inlined]
   [11] macro expansion
      @ C:\Users\sam.urmy\.julia\packages\StaticArrays\dTwvg\src\broadcast.jl:140 [inlined]
   [12] __broadcast
      @ C:\Users\sam.urmy\.julia\packages\StaticArrays\dTwvg\src\broadcast.jl:128 [inlined]
   [13] _broadcast
      @ C:\Users\sam.urmy\.julia\packages\StaticArrays\dTwvg\src\broadcast.jl:124 [inlined]
   [14] copy
      @ C:\Users\sam.urmy\.julia\packages\StaticArrays\dTwvg\src\broadcast.jl:62 [inlined]
   [15] materialize
      @ .\broadcast.jl:873 [inlined]
   [16] _eval_function(m::FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}, f::FiniteDifferences.var"#74#76"{Int64, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}}, x::Float64, step::Float64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\methods.jl:249
   [17] _estimate_magnitudes(m::FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}, f::FiniteDifferences.var"#74#76"{Int64, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}}, x::Float64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\methods.jl:378
   [18] estimate_step(m::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::FiniteDifferences.var"#74#76"{Int64, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}}, x::Float64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\methods.jl:365
   [19] (::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}})(f::FiniteDifferences.var"#74#76"{Int64, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}}, x::Float64)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\methods.jl:193
   [20] #73
      @ C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:16 [inlined]
   [21] iterate
      @ .\generator.jl:47 [inlined]
   [22] _collect(c::Base.OneTo{Int64}, itr::Base.Generator{Base.OneTo{Int64}, FiniteDifferences.var"#73#75"{FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, Vector{Float64}}}, #unused#::Base.EltypeUnknown, isz::Base.HasShape{1})
      @ Base .\array.jl:802
   [23] collect_similar
      @ .\array.jl:711 [inlined]
   [24] map
      @ .\abstractarray.jl:3263 [inlined]
   [25] jacobian(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::FiniteDifferences.var"#91#92"{FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64, Int64}, Tuple{Int64, Int64, Int64}, Tuple{FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}, FiniteDifferences.var"#structtype_from_vec#29"{Pair{Int64, Vector{Float64}}, FiniteDifferences.var"#Tuple_from_vec#53"{Tuple{Int64, Int64}, Tuple{Int64, Int64}, Tuple{typeof(identity), typeof(identity)}}, Tuple{FiniteDifferences.var"#Real_from_vec#20", typeof(identity)}}}}}, x::Vector{Float64}; len::Nothing)
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:15
   [26] jacobian(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::Function, x::Vector{Float64})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:9
   [27] _j′vp(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::Function, ȳ::Vector{Float64}, x::Vector{Float64})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:84
   [28] j′vp(fdm::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, f::FiniteDifferences.var"#93#94"{ChainRulesTestUtils.var"#fnew#53"{ChainRulesTestUtils.var"#call#63"{NamedTuple{(), Tuple{}}}, Tuple{typeof(spdiagm), Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}}, NTuple{4, Bool}}}, ȳ::SparseMatrixCSC{Float64, Int64}, x::Tuple{Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}, Pair{Int64, Vector{Float64}}})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:77
   [29] j′vp(::FiniteDifferences.AdaptedFiniteDifferenceMethod{5, 1, FiniteDifferences.UnadaptedFiniteDifferenceMethod{7, 5}}, ::Function, ::SparseMatrixCSC{Float64, Int64}, ::Pair{Int64, Vector{Float64}}, ::Pair{Int64, Vector{Float64}}, ::Vararg{Pair{Int64, Vector{Float64}}})
      @ FiniteDifferences C:\Users\sam.urmy\.julia\packages\FiniteDifferences\BSR84\src\grad.jl:80
   [30] _make_j′vp_call(fdm::Any, f::Any, ȳ::Any, xs::Any, ignores::Any)
      @ ChainRulesTestUtils C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\finite_difference_calls.jl:51
   [31] macro expansion
      @ C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:233 [inlined]
   [32] macro expansion
      @ C:\Users\sam.urmy\.julia\juliaup\julia-1.9.3+0.x64.w64.mingw32\share\julia\stdlib\v1.9\Test\src\Test.jl:1498 [inlined]
   [33] test_rrule(::RuleConfig, ::Any, ::Any, ::Vararg{Any}; output_tangent::Any, check_thunked_output_tangent::Any, fdm::Any, rrule_f::Any, check_inferred::Bool, fkwargs::NamedTuple, rtol::Real, atol::Real, testset_name::Any, kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
      @ ChainRulesTestUtils C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:205
   [34] test_rrule(::RuleConfig, ::Any, ::Any, ::Vararg{Any})
      @ ChainRulesTestUtils C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:181
   [35] test_rrule(::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
      @ ChainRulesTestUtils C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:178
   [36] test_rrule(::Any, ::Any, ::Any, ::Vararg{Any})
      @ ChainRulesTestUtils C:\Users\sam.urmy\.julia\packages\ChainRulesTestUtils\C9L2i\src\testers.jl:177
   [37] macro expansion
      @ C:\Users\sam.urmy\.julia\dev\ChainRules\test\rulesets\SparseArrays\sparsematrix.jl:31 [inlined]
   [38] macro expansion
      @ C:\Users\sam.urmy\.julia\juliaup\julia-1.9.3+0.x64.w64.mingw32\share\julia\stdlib\v1.9\Test\src\Test.jl:1498 [inlined]
   [39] macro expansion
      @ C:\Users\sam.urmy\.julia\dev\ChainRules\test\rulesets\SparseArrays\sparsematrix.jl:24 [inlined]
   [40] macro expansion
      @ C:\Users\sam.urmy\.julia\juliaup\julia-1.9.3+0.x64.w64.mingw32\share\julia\stdlib\v1.9\Test\src\Test.jl:1498 [inlined]
   [41] top-level scope
      @ C:\Users\sam.urmy\.julia\dev\ChainRules\test\rulesets\SparseArrays\sparsematrix.jl:23
   [42] include(fname::String)
      @ Base.MainInclude .\client.jl:478
   [43] macro expansion
      @ C:\Users\sam.urmy\.julia\dev\ChainRules\test\runtests.jl:86 [inlined]
   [44] macro expansion
      @ C:\Users\sam.urmy\.julia\juliaup\julia-1.9.3+0.x64.w64.mingw32\share\julia\stdlib\v1.9\Test\src\Test.jl:1498 [inlined]
   [45] top-level scope
      @ C:\Users\sam.urmy\.julia\dev\ChainRules\test\runtests.jl:50
   [46] include(fname::String)
      @ Base.MainInclude .\client.jl:478
   [47] top-level scope
      @ none:6
   [48] eval
      @ .\boot.jl:370 [inlined]
   [49] exec_options(opts::Base.JLOptions)
      @ Base .\client.jl:280
   [50] _start()
      @ Base .\client.jl:522
```
</details>

Reply:
Fair enough,